### PR TITLE
8288966: Better handle very spiky promotion in G1

### PR DESCRIPTION
**Pull request by tschatzl**
Hi all,

  please review this change that implements "boosting" the PLAB sizes after every "significant" amount of refills.

This removes pause time spikes during garbage collection if the PLAB sizing algorithm has not picked up a good value yet (at startup) or the load is very spiky (e.g. SPECjbb2015).

This works by defining a "tolerated amount of refills", after every time this threshold is met we double the PLAB. The reason for not doing this doubling every time is waste - on tight(er) heaps this can waste a lot of memory for very small allocations, causing more GCs and effective slowdowns.

So this threshold is based on the number of expected refills (per thread); allowing a little more than these expected refills (and discounting the initial PLAB refill).
This boosting is also only active with `ResizePLAB` - it's on by default anyway with G1.

The CR contains a graph showing massive pause time spikes on SPECjbb2015 that are fixed by this, as otherwise during these spikes there would be 500k+ PLAB refills that in particular seem to be very harmful on aarch64 (the machines seem to be very problematic wrt to contention on a single atomic counter with lots of threads).

Overall performance (of SPECjbb2015, on aarch64) is slightly better with that change, and in general also improves startup pauses. x64 isn't really affected either way.

I'm aware that now `G1PLABAllocator` has lots of members that are arrays of the same size (the number of generations). I want to fix that to an array of structs in a separate change.

Thanks,
  Thomas

Testing: tier1-5

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8288966](https://bugs.openjdk.org/browse/JDK-8288966): Better handle very spiky promotion in G1

### Reviewers
 * [Ivan Walulya](https://openjdk.org/census#iwalulya) (@walulyai - **Reviewer**) ⚠️ Review applies to [7a819e23](https://git.openjdk.org/jdk/pull/9726/files/7a819e23756890af9946949832578f7ed46cf168)
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to [ce2701cb](https://git.openjdk.org/jdk/pull/9726/files/ce2701cbc581a3085eec02927f5fc5260167e48b)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9726/head:pull/9726` \
`$ git checkout pull/9726`

Update a local copy of the PR: \
`$ git checkout pull/9726` \
`$ git pull https://git.openjdk.org/jdk pull/9726/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9726`

View PR using the GUI difftool: \
`$ git pr show -t 9726`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9726.diff">https://git.openjdk.org/jdk/pull/9726.diff</a>

</details>

### CODE REVIEW

**Comment by tschatzl** [comment]
Filed [JDK-8291912](https://bugs.openjdk.org/browse/JDK-8291912) for the mentioned refactoring.

**Review by walulyai — APPROVED** [review]
Lgtm!

File: src/hotspot/share/gc/g1/g1Allocator.cpp [hunk 1]
```diff
@@ -296,6 +296,24 @@ HeapWord* G1Allocator::old_attempt_allocation(size_t min_word_size,
...
+    // were reached.
+    // E.g. due to limitation of PLAB size to non-humongous objects and region boundaries
+    // a thread may experience more refills than expected. Keeping the PLAB waste low
+    // is the main goal, so being a bit conservative is better.
+    double const PadFactor = 1.5;
+    _tolerated_refills = MAX2(ExpectedNumberOfRefills, 1.0) * PadFactor;
+  } else {
+    // Make the tolerated refills a huge number; -2 because we add one to this
+    // value later and it would overflow otherwise.
+    _tolerated_refills = (size_t)-2;
```
Comment (kimbarrett) [inline src/hotspot/share/gc/g1/g1Allocator.cpp:314]: `SIZE_MAX - 1` would be more obvious.

**Review by kimbarrett — APPROVED** [review]
Looks good.

File: src/hotspot/share/gc/g1/g1ParScanThreadState.hpp [hunk 2]
```diff
@@ -152,7 +152,7 @@ class G1ParScanThreadState : public CHeapObj<mtGC> {
 
   // Pass locally gathered statistics to global state. Returns the total number of
   // HeapWords copied.
-  size_t flush(size_t* surviving_young_words);
+  size_t flush(size_t* surviving_young_words, uint num_workers);
```
[the remark this answers is not in the document]
Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1ParScanThreadState.hpp:155]: I'll do a rename CR for that.

**Comment by tschatzl** [comment]
Thanks @kimbarrett @walulyai for your reviews

/integrate

**Pull request closed without merging**